### DISCUSSION TITLE
how could i successfully parse configuration with "remove" attribute with lyd_parse_data_mem api?

### DISCUSSION THREAD

**Discussion by lyzliyuzhi91** [CESNET/libyang#2026] [Q&A] [answered]
```
module a {
    yang-version 1.1;
    namespace "urn:huawei:xxx:unit-test:a";
    prefix a;

    container c1 {
        leaf num-leaf {
            type uint32;
        }
    }
}
```

```
    source = "{\"a:c1\":{\"num-leaf\":\"\",\"@num-leaf\":{\"ietf-netconf:operation\":\"remove\"}}}";
    uint32_t parse_opts = LYD_PARSE_STRICT | LYD_PARSE_ONLY | LYD_PARSE_NO_STATE;
    err = lyd_parse_data_mem(ctx, source, LYD_JSON, parse_opts, LYD_VALIDATE_PRESENT, &out);
    ASSERT_EQ(err, LY_SUCCESS);
```

But `err` return `LY_EVALID` .
Thanks.

**Comment by michalvasko** [comment]
Look at the error message, if you do not want to enable logging then with `ly_last_errmsg()`, for example. Only one problem I can think of based on this brief description, have you loaded the `ietf-netconf` YANG into the context?

**Reply by jiayinhaoran** [reply]
I also find the same issue that when leaf data node have no value(ex: remove this leaf or query this leaf), lyd_parse_data_mem will return error and the error normally is relatived with node type(ex:data is: {\"uint32-data\":\"\",\"@interval-time\":{\"ietf-netconf:operation\":\"remove\"}}  and error is Invalid non-number-encoded uint32 value \"\") or type limit(ex: query leaf value is: {\"sys-name\":\"\"} and error is : Unsatisfied length - string "" length is not allowed.)

How to correctlly parse a query operation or remove operation, thanks @michalvasko

**Reply by lyzliyuzhi91** [reply]
Error message:
libyang[0]: (Path:Data location "/a:c1/num-leaf", line number 1.)Invalid non-number-encoded uint32 value "".                                                                                                     

> have you loaded the `ietf-netconf` YANG into the context?
Yes, I have loaded the `ietf-netconf` module.

**Reply by lyzliyuzhi91** [reply]
What i want to do is wanting to enforce some validation for the `<config>` node in `<edit-config>` , for example, the newly created/merged node's value, but I think the value of nodes to be deleted should not be validated?

**Reply by michalvasko** [reply]
Right, I have missed that the value you are using is invalid. If you want to parse it, you have to use `LYD_PARSE_OPAQ` but such nodes will not validate. That should not matter, though, so just get rid of `LYD_VALIDATE_PRESENT`, there is nothing to validate in an edit config data anyway.

**Reply by lyzliyuzhi91** [reply]
But once I use LYD_PARSE_OPAQ, I cannot parse the value of newly create/merge right?
>  there is nothing to validate in an edit config data anyway

Why? I want to check the validation for the configuration. For example, if a string-type leaf node `a` is restricted by `length  8..20`, and the user configure `a` with value `"xxx"`. And once I get the edit-config rpc message, I just want to get a result that value of `a` is incorrect by libyang's validation function.

**Reply by michalvasko** [reply, marked as answer]
> But once I use LYD_PARSE_OPAQ, I cannot parse the value of newly create/merge right?

Do not understand what you mean, what the flag does is parses invalid nodes as opaque nodes, valid nodes are parsed normally.

> Why? I want to check the validation for the configuration.

You are not supposed to be validating the edit config (because it can be invalid as seen in this case) but rather the configuration after the edit is applied.
